File: projects/pyflowcl/code/test_libtorch/conv_FD.py
import numpy as np
import torch
import time

import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
plt.rcParams.update({'font.size': 16})

from matplotlib import cm,rc


# Dimensions
Nxi  = 512
Neta = 512

# Convective velociites
cx = 1.0
cy = 1.0

# Stopping condition
nper = 1
dt   = 1e-3
Nsteps = 2000


# --------------------------------------------------------------
# Grid
#  Periodic, so N+1 point is redundant
eta_min = 0.25
eta_max = 4.0

xi_grid  = torch.linspace(0,2*np.pi,Nxi+1)         # xi = theta
eta_grid = torch.linspace(eta_min,eta_max,Neta+1)  # eta = r

d_xi  = 2*np.pi/float(Nxi)
d_eta = (eta_max - eta_min)/float(Neta)

# Definition of curvilinear x,y coordinates
def get_transform(xi,eta):
    x = eta * np.cos(xi)
    y = eta * np.sin(xi)
    return x,y

# Curvilinear grid metrics
#   Analytic derivatives of x,y with respect to xi,eta
def get_metrics(xi,eta):
    x_xi  = -eta * torch.sin(xi)
    y_xi  =  eta * torch.cos(xi)
    x_eta = torch.cos(xi)
    y_eta = torch.sin(xi)

    # Grid Jacobian
    Jac = x_xi * y_eta - x_eta * y_xi
    
    return x_xi,y_xi,x_eta,y_eta,Jac


# Grid metrics differ for each x,y. Compute using meshgrid.
Xi,Eta = torch.meshgrid(xi_grid[:-1],eta_grid[:-1])
x_xi,y_xi,x_eta,y_eta,Jac = get_metrics(Xi,Eta)

inv_Jac = 1.0/Jac

# x,y meshgrids for plotting
X,Y = get_transform(Xi,Eta)


# --------------------------------------------------------------
# Gaussian initial condition
U = torch.zeros((Nxi,Neta), dtype=torch.float64)
xloc = -2
yloc = 0
for i,xi in enumerate(xi_grid[:-1]):
    for j,eta in enumerate(eta_grid[:-1]):
        x,y = get_transform(xi,eta)
        if (abs(x-xloc)<=1.0 and abs(y-yloc)<=1.0):
            U[i,j] = ( np.exp(1.0/((x-xloc)**2 - 1.0)) * \
                       np.exp(1.0/((y-yloc)**2 - 1.0)) )

def get_energy(u):
    return torch.sum(u**2).numpy()

energy_IC = get_energy(U)
print(energy_IC)


# --------------------------------------------------------------
# RHS function
def RHS(u):
    E = ( cx * y_eta - cy * x_eta ) * u
    F = ( cy * x_xi  - cx * y_xi  ) * u
    
    xl = torch.cat((E[-1:,:],E[:-1,:]), dim=0)
    xr = torch.cat((E[1:,:], E[:1,:])  , dim=0)
    xi_term = -( xr - xl ) / (2.0*d_xi)

    ydn = torch.cat((F[:,-1:],F[:,:-1]), dim=1)
    yup = torch.cat((F[:,1:], F[:,:1]) , dim=1)
    eta_term = -( yup - ydn ) / (2.0*d_eta)

    del xl,xr,ydn,yup
    return inv_Jac * (xi_term + eta_term)


# --------------------------------------------------------------
# Solution plotter
def plot_sol(u,i):
    plt.pcolormesh(X,Y,u, vmin=-0.01,vmax=0.14)
    plt.colorbar()
    plt.savefig('Output/sol_{:06d}.png'.format(i))
    plt.close()

def monitor(n,t,u):
    print('{:6d}  {:10.5e}  {:10.5e}'.format(n, t, get_energy(u)/energy_IC))
    #plot_sol(u.numpy(),n)

    
# --------------------------------------------------------------
# Time stepping
t = 0
time1 = time.time()

for n in range(Nsteps):
    
    # Monitoring
    if (n%50==0): monitor(n,t,U)
        
    # RK4
    k1 = dt * RHS( U )
    k2 = dt * RHS( U + 0.5*k1 )
    k3 = dt * RHS( U + 0.5*k2 )
    k4 = dt * RHS( U + k3 )

    U += (k1 + 2.0*k2 + 2.0*k3 + k4)/6.0
    t += dt

# Done time-stepping

monitor(n+1,t,U)

time2 = time.time()
print('Done solving, elapsed={:9.5f}'.format(time2-time1))

# Final plot
#plot_sol(U.numpy(),Nsteps)

File: projects/pyflowcl/code/test_libtorch/conv_FD.cpp
#include <torch/torch.h>
#include <iostream>
using namespace torch::indexing;

#define pi 3.141592653589793238462643383279502884


void RHS( torch::Tensor u,
	  torch::Tensor udot,
	  torch::Tensor x_xi,
	  torch::Tensor y_xi,
	  torch::Tensor x_eta,
	  torch::Tensor y_eta,
	  torch::Tensor inv_Jac,
	  double d_xi, double d_eta) {

  // Grid size
  int Nx = 512;
  int Ny = 512;
  
  // Convective velocities
  double cx = 1.0;
  double cy = 1.0;
  
  torch::Tensor E = ( cx * y_eta - cy * x_eta ) * u;

  torch::Tensor xl = torch::cat({ E.index({Slice(Nx-1,Nx), Slice(0,Ny)}),
				  E.index({Slice(0, Nx-1), Slice(0,Ny)}) }, 0);

  torch::Tensor xr = torch::cat({ E.index({Slice(1,Nx), Slice(0,Ny)}),
				  E.index({Slice(0, 1), Slice(0,Ny)}) }, 0);

  torch::Tensor xi_term = -( xr - xl ) / (2.0*d_xi);

  
  E = ( cy * x_xi  - cx * y_xi  ) * u;

  xl = torch::cat({ E.index({ Slice(0,Nx), Slice(Ny-1,Ny) }) ,
		    E.index({ Slice(0,Nx), Slice(0, Ny-1) }) }, 1);

  xr = torch::cat({ E.index({ Slice(0,Nx), Slice(1,Ny) }) ,
		    E.index({ Slice(0,Nx), Slice(0, 1) }) }, 1);

  torch::Tensor eta_term = -( xr - xl ) / (2.0*d_eta);

  udot = inv_Jac * (xi_term + eta_term);
}


double get_energy( torch::Tensor u ) {
  return torch::sum( u*u ).item().to<double>();
}
  

int main() {
  //torch::Tensor tensor = torch::rand({2, 3});
  //std::cout << tensor << std::endl;

  // Grid size
  int Nx = 512;
  int Ny = 512;

  // Sopping condition
  int Nsteps = 2000;
  double dt  = 1e-3;

  auto options =
    torch::TensorOptions()
    .dtype(torch::kFloat64)
    .layout(torch::kStrided);
    //.device(torch::kCUDA, 1)
    //.requires_grad(true);

  // ----------------------------------------------------
  // Grid
  double eta_min = 0.25;
  double eta_max = 4.0;

  torch::Tensor xi_grid  = torch::linspace(0, 2.0*pi, Nx+1, options);
  torch::Tensor eta_grid = torch::linspace(eta_min, eta_max, Ny+1, options);

  double d_xi  = 2.0*pi/(float) Nx;
  double d_eta = (eta_max - eta_min)/(float) Ny;

  std::vector<torch::Tensor> Grid = torch::meshgrid({ xi_grid.index({Slice(0,Nx)}),
						      eta_grid.index({Slice(0,Ny)}) });
  torch::Tensor Xi  = Grid[0];
  torch::Tensor Eta = Grid[1];
  
  torch::Tensor x_xi  = -Eta * torch::sin(Xi);
  torch::Tensor y_xi  = -Eta * torch::cos(Xi);
  torch::Tensor x_eta = torch::cos(Xi);
  torch::Tensor y_eta = torch::sin(Xi);

  torch::Tensor inv_Jac = 1.0/(x_xi * y_eta - x_eta * y_xi);

  torch::Tensor X = Eta * torch::cos(Xi);
  torch::Tensor Y = Eta * torch::sin(Xi);
  //std::cout << Y.dtype() << std::endl; // double

  // ----------------------------------------------------
  // Initial condition
  torch::Tensor U = torch::zeros({Nx,Ny}, options);
  //std::cout << U.dtype() << std::endl; // double

  double xloc = -2.0;
  double yloc = 0.0;
  
  for (int i=0; i<Nx; i++) {
    for (int j=0; j<Ny; j++) {
      double x = X[i][j].item().to<double>();
      double y = Y[i][j].item().to<double>();
      if (std::abs(x - xloc)<=1.0 &&
	  std::abs(y - yloc)<=1.0) {
	U[i][j] = ( std::exp(1.0 / (std::pow(x - xloc,2) - 1.0)) *
		    std::exp(1.0 / (std::pow(y - yloc,2) - 1.0)) );
      }
    }
  }

  double energy_IC = get_energy(U);

  // ----------------------------------------------------
  // Time stepping
  double t = 0.0;
  torch::Tensor K1 = torch::zeros({Nx,Ny}, options);
  torch::Tensor K2 = torch::zeros({Nx,Ny}, options);
  torch::Tensor K3 = torch::zeros({Nx,Ny}, options);
  torch::Tensor K4 = torch::zeros({Nx,Ny}, options);
  
  for (int n=0; n<Nsteps; n++) {

    if (n%50==0) {std::cout << n << '\t'
			    << get_energy(U) / energy_IC
			    << std::endl;}

    // RK4
    RHS( U, K1, x_xi, y_xi, x_eta, y_eta, inv_Jac, d_xi, d_eta );
    RHS( U + 0.5*dt*K1, K2, x_xi, y_xi, x_eta, y_eta, inv_Jac, d_xi, d_eta );
    RHS( U + 0.5*dt*K2, K3, x_xi, y_xi, x_eta, y_eta, inv_Jac, d_xi, d_eta );
    RHS( U + dt*K3, K4, x_xi, y_xi, x_eta, y_eta, inv_Jac, d_xi, d_eta );

    U = U + dt*(K1 + 2.0*(K2 + K3) + K4)/6.0;

    t += dt;
    
  }

}
